From 797b5166783cda0886038ffb22f5386b9363a961 Mon Sep 17 00:00:00 2001
From: Robert Marko <robimarko@gmail.com>
Date: Sun, 22 Jan 2023 22:08:27 +0100
Subject: [PATCH 3/5] cryptoapi: v2.0: remove setting crypto_ahash_type for
 newer kernels

Upstream has stopped exporting crypto_ahash_type and removed setting it
on ahash algos since v4.19 as its easily identifiable by the struct type
and its being set in the core directly, so lets do the same.

Signed-off-by: Robert Marko <robimarko@gmail.com>
---
 cryptoapi/v2.0/nss_cryptoapi.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

--- a/cryptoapi/v2.0/nss_cryptoapi.c
+++ b/cryptoapi/v2.0/nss_cryptoapi.c
@@ -1495,7 +1495,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = MD5_HMAC_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1521,7 +1523,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA1_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1547,7 +1551,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA224_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1573,7 +1579,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA256_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1599,7 +1607,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA384_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1625,7 +1635,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA512_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1655,7 +1667,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = MD5_HMAC_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1681,7 +1695,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA1_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1707,7 +1723,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA224_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1733,7 +1751,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA256_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1759,7 +1779,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA384_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
@@ -1785,7 +1807,9 @@ static struct ahash_alg cryptoapi_ahash_
 				.cra_blocksize   = SHA512_BLOCK_SIZE,
 				.cra_ctxsize     = sizeof(struct nss_cryptoapi_ctx),
 				.cra_alignmask   = 0,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0)
 				.cra_type        = &crypto_ahash_type,
+#endif
 				.cra_module      = THIS_MODULE,
 				.cra_init        = nss_cryptoapi_ahash_cra_init,
 				.cra_exit        = nss_cryptoapi_ahash_cra_exit,
